llvm/lib/Transforms/Scalar/EarlyCSE.cpp

```diff
@@ -1632,6 +1639,18 @@ bool EarlyCSE::processNode(DomTreeNode *Node) {
           LLVM_DEBUG(dbgs() << "Skipping due to debug counter\n");
           continue;
         }
+        // We probably don't need a complete itersection of attrs between
+        // InVal.first and Inst.
+        // We only CSE readonly functions that have the same memory state and
+        // its not clear that any non-return *callsite* attribute can create
+        // side-effects. Likewise this implies when checking equality of
+        // callsite for CSEing, we can probably ignore all non-return attrs.
```

Comment:
This is not right, there is really no distinction between return and parameter attributes here. Just move the attribute to the parameter and have the function return it.

There *is* a distinction between attributes that imply poison and those that imply UB, in that the latter can be unconditionally preserved from the first call-site, while the former need intersection. But just intersecting everything is conservatively correct.

Reply:
Your argument is something like:

```
declare void @biz(ptr, ptr)
define ptr @bar(ptr %p) {
   ret ptr %p
}

define void @buz(ptr %p) {
  %p0 = call ptr @bar(ptr nonnull %p) readnone
  %p1 = call ptr @bar(ptr %p) readnone
  call void @biz(ptr %p0, ptr %p1)
}
```

Where we later propagate the `nonnull` to the return attr of `bar`?

Reply:
Cleared up the comment.